### DISCUSSION TITLE
NetBeans 13, 12, ..., fails to build NetBeans Platform Application with Maven

### DISCUSSION THREAD

**Discussion by Addlai** [apache/netbeans#3846] [General]
### Apache NetBeans version

Apache NetBeans 13

### What happened

1. I have been unable to build any Maven NetBeans Application project, including both an empty project created from 
File/New Project...
or any project I've tried compiling from downloaded samples, projects on GitHub, etc.

2. No previous version of the NetBeans IDE I've tried (12.6, 11.4, 10, 9) seems to be able to compile a NetBeans application either. I've seen references to building NetBeans Applications with IDE version 8.2. However, the link to the NetBeans IDE version 8.2 is broken on the download page, so I've been unable to test it.

3. I've found few, if any, online discussions pertaining to the NetBeans Platform.

4. The documentation and tutorials for the NetBeans Platform on the Apache NetBeans web site are out of date and filled with broken links. For instance, all 10 tutorial videos on https://netbeans.apache.org/tutorials/nbm-10-top-apis.html (Top 10 NetBeans APIs) seem to be broken links and internet searches turn up nothing.

5. It appears therefore that the NetBeans Platform is no longer supported, at least for projects outside of the NetBeans IDE itself. Yet, the menu items and web pages are still present. I have spent quite a lot of time exploring this issue. I strongly suggest that, if it is the case, the NetBeans team announce that support for the NetBeans Platform has been terminated, and the seeming functionality be removed from the NetBeans IDE so that other developers do not lose time attempting to use it. 

6. Instead, I suggest that the NetBeans team restore the link to IDE version 8.2, and the documentation such as it existed for the most recent version of the IDE that still supported the NetBeans Platform, so that anyone who chooses to use the NetBeans Platform with an older IDE and JDK has a way of doing so.

### How to reproduce

File/New Project...
Categories: Java with Maven
Projects: NetBeans Application
Click Next >
(Accept defaults)
Click Next >
(Accept defaults ... RELEASE90 ...)
Click Finish

The project build fails with 

[ERROR] Some problems were encountered while processing the POMs:
Child module /Users/***/mavenproject1/branding of /Users/***/mavenproject1/pom.xml does not exist @ 
Child module /Users/***/mavenproject1/application of /Users/***/mavenproject1/pom.xml does not exist @ 
 @ 
The build could not read 1 project -> [Help 1]

### Did this work correctly in an earlier version?

No

### Operating System

MacOS Monterry 12.3

### JDK

I have tried JDK 8, 11, and 17.

### Apache NetBeans packaging

Apache NetBeans provided installer

### Anything else

With other combinations of the IDE, JDK, etc., I get different errors. For instance, 
Failed to execute goal org.apache.maven.plugins:maven-archetype-plugin:3.1.2:generate (default-cli) on project XXX: Unable to add module to the current project as it is not of packaging type 'pom' -> [Help 1]

Note: XXX is not the name of the project, which is actually mavenproject1-parent.

### Are you willing to submit a pull request?

No

### Code of Conduct

Yes

**Comment by matthiasblaesing** [comment]
Too much text here. "What happened" makes assumptions and jumps to conclusions that are either illogical or wrong and should not be there. I just followed the "How to reproduce" steps and they work as they should, i.e. they create the three skeleton projects (pom packaging parent, application, branding). "Clean and build" builds, so at this point it is "Works for me".

This is NetBeans build from master a few days ago with the Ubuntu OpenJDK.

For the state of the documentation, there is a well visibile note on top of that page:

> This tutorial needs a review.

So you are warned and you could be the one rebuilding the documentation. This is a project done by volunteers, so if you want something done, consider doing it yourself.

**Reply by Addlai** [reply]
Perhaps you could be specific about where I jumped to conclusions that are illogical or wrong. It would help a crazy man like me to know where I've strayed so far.

Perhaps it's hasty of me to draw any conclusions from the fact that not 1, not 2, ... but in fact all 10 links in the recommended documentation entry point (https://netbeans.apache.org/tutorials/nbm-10-top-apis.html) are broken links to videos created more than a decade ago? And that the page's GitHub repository has the Issues feature turned off, so I can't report the problem there. And that nobody here seems overly concerned about this situation?

This platform took a decade or two to create. It's the only comprehensive RCP with desktop Swing applications. Do you want people to use it or not? Every single link is broken. And there's nowhere else to go. And nobody seems alarmed on the team. Who is being illogical here?

Perhaps putting any significance in the that that creating an empty Maven NetBeans application, with every default setting, fails? Or that every attempt to build every single Maven NetBeans application I've been able to find also fails?

Or is it one of my other illogical leaps that you're referring to?

**Comment by mbien** [comment]
i can't reproduce this. I just created a NB platform application based on 13 and it worked fine.

Build the parent project first so that maven downloads everything, after that you should be able to run the app project. JDK 17 will require a few additional flags to configure the java module system but it should start on JDK 11 out-of-the-box.

**Reply by Addlai** [reply]
I'm not sure I understand what you mean. Is JDK 17 not supported out of the box? If not, why not? And why is this not stated in the release notes if it is the case. I'm hoping to move a recent code base developed on JDK 17 to the NetBeans Platform.

Or do you mean by "require a few additional flags to configure" that the user is expected to reconfigure something in order to use JDK 17. I am new to the platform, and I can't locate much of the tutorial information. How would I be expected to know what flags to set?

And you aren't even testing it on the same operating system as me. And it's not clear from what you're saying that you're testing it on a clean machine.

Has it ever happened that a problem appears on an end user's machine running one operating system, but not on one of the developer's own machines running a different operating system?

**Reply by mbien** [reply]
java devs which use modern JDKs should be aware of JEP 403. If they aren't yet, looking at the exception msg should clearly point out the issue - it is clearly stating which JDK module can't be accessed, even down to the package.

The list of --add-opens flags is highly dependent on the actual dependencies (not just NB dependencies!) you will be using in your platform application and also from your own code (of course). Providing a perfect list will be hard, what we likely will do for NB 14 is to provide a full list of initial flags which are required to run the full IDE, very few platform users will want to do that though. This is might be convenient but should not be used in production.

Additionally: the build system in question is maven. Users can configure plugins and all kind of build goals in it where they have to deal with those flags anyway. This is standard java run configuration, not unique to the NB platform. This is IDE independent in fact, all projects should be buildable outside of IDEs. You can use the netbeans platform maven archetype from the command line to setup a new project.

So yes, the NB platform can run on JDK 17, if the JVM is configured correctly. An example for that is the NetBeans 13 IDE itself which has the config in etc/netbeans.conf and runs on JDK 17 just fine.

if you want to help, test #3851 (still WIP), or help improving the documentation, e.g here https://github.com/apache/netbeans-website/

**Comment by neilcsmith-net** [comment]
Let's move this to discussions.  There may be issues to address that stem from this, but it's too broad as is to be useful here.

The platform is still supported, and actively used by multiple people involved in the NetBeans project.

@ebarboni may have some thoughts on the Maven tooling?  I generally use Ant still for RCP.

We have no way (as Apache) of restoring links to NetBeans 8.2 - it wasn't an Apache project at that time!

**Reply by Addlai** [reply]
I don't understand why this is moved to discussions. In the "How to Reproduce" I stated how to do this with 3 mouse clicks. Your comments make it sound like you aren't concerned with a show stopping bug, that prevents creation of ANY Maven NetBeans applications. A similar problem (Failed to create maven enterprise application #3723) is still open.

I'd also like to point out, that if you are still using Ant, you're not reproducing my situation. I never stated there was a bug with creating an application using Ant. It seems likely in fact that what's going wrong has to do with Maven, possibly nbm-maven-plugin.

The other things I state are to give context which I think is relevant. This doesn't appear to be an isolated problem. It's affecting every single Maven NetBeans application I've tried to build. And every single version of NetBeans that's available on the Apache site seems to be affected. If you're not alarmed, let me be alarmed for you. This is a show stopping bug!

If someone assigned the bug doesn't want to read my additional comments, they don't need to.

However, I did take the opportunity to ask if the platform was still being supported, because it's not clear at all to an outsider whether it is or isn't. There's a lot of confusion on the web site, with a very large number of pages with broken links. I've never seen anything like that in a project that's active. 

There was a switch from Jira to GitHub, but the Issues feature is turned off on GitHub, so you can't report these problems there.

**Reply by neilcsmith-net** [reply]
@Addlai it was moved to discussions because there are multiple issues at play here that need splitting in to actionable tasks.

When I said "generally use Ant" it doesn't mean that I don't test with Maven.  Using NB13 and `RELEASE130` (which are the modules from NB13) generates fine here.  Part of this is going to be improving UI and fixing docs.

**Comment by ebarboni** [comment]
For me this works well on the windows os I have no access to macos to reproduce that. Would be nice to know if some IDE log (MENU -> view -> IDE LOG) are displayed maybe folder creation limitation on the os because to me it seems that it fails in generating the folder from the archetype itself and that later maven clean install fail at low project level.

The only "issue" may be to bump nbm-maven plugin to 4.7 and netbeans release version to a release not containg the pack2000 artefacts like the  RELEASE130 (RELEASE90 to old) to support jdk 17 and the pack2000 removal and compiler maven plugin also that may be old. Harness delivered is also a bit old but still works.

**Reply by Addlai** [reply]
Windows != MacOS. Developer machine != user machine. 

I'm not sure what is mean't by "folder creation limitation"? Even if MacOS had a limitation in the creation of folders ... write once run everywhere.

As far as the NetBeans platform version, on a fresh installation, RELEASE110 is the only choice offered by default for NetBeans IDE 13. (RELEASE90 was offered under 12.6.). If there are reasons to choose one release over another, I don't know how I'd be expected to know that. 

If I need to "bump nbm-maven-plugin to 4.7 ..." One of the many things I tried, was replacing the 4.3 dependency in the pom file by 4.7. That didn't fix the problem, so I put it back to 4.3.

**Comment by Addlai** [comment]
One more point. I took the trouble of wiping the hard drive of a MacBook pro I borrowed from my employer. On a completely clean machine, I installed MacOS Catalina 10.15.7, then JDK 17 (amazon corretto 17.0.2.8.1), then a freshly downloaded copy of NetBeans 13. No other software was installed. Using the same 3 mouse clicks, I reproduced the build failure.

Failed to execute goal org.apache.netbeans.utilities:nbm-maven-plugin:4.3:cluster-app (default-cluster-app: Execution default-cluster-app of goal org.apache.netbeans.utilities:nbm-maven-plugin: ...

**Comment by geertjanw** [comment]
You’re going to need to turn down your frustration levels significantly here. We’re all volunteers, all working here for free — for example, has it occurred to you that if you find links to broken tutorials or documentation that you yourself as a user of an open source project are responsible for replacing those broken links and tutorials and providing new content, on your own or in collaboration with others?

**Comment by geertjanw** [comment]
Also, are you using the NetBeans Platform in a commercial setting, i.e., are you or is your employer making money from the NetBeans Platform? We’d welcome you or your employer paying for someone to fix each and every bug that bothers you. How does that sound?

**Reply by Addlai** [reply]
The last thing I want to spend my time on right now is being here. I'm trying to play a role in bringing attention to problems that may well be hindering NetBeans.

I will be paid by my employer for the project, which is not my main job, an amount that averages to less than minimum wage. They have signed a legal agreement with me that prevents them from commercializing the software. I have milestones with deadlines. If they aren't met, I won't be paid at all.

Nevertheless, my employer could probably be persuaded to pay you something reasonable to fix your IDE, if that's what you need. It seems odd to pay the NetBeans IDE team to make their own IDE compile for their own platform, but if that's what it will take, I'll bring it to them.

At some point, down the road, it's possible that when I know more, I could contribute fixes myself. Right now, it would just be nice to have access to the documentation.

**Comment by geertjanw** [comment]
It seems odd to contribute time and money to an open source project?

It seems odd that you think people will work for free to support a commercial project. It won’t be me anyway.

**Reply by Addlai** [reply]
OK, don't fix NetBeans. You win.

My employer, which is a community college, hoping to use the software to manage developmental educational needs, will get by.

You win.

**Reply by geertjanw** [reply]
So… not a commercial organization at all, then. I don’t understand why not just say that up front? :-)

And how does anyone “win”? So much to be confused about here.

Anyway, will do my very best to reproduce and help!

**Comment by sdedic** [comment]
Wow, what a heated discussion here. FWIW, I've just tried the same with NB13 release and ...
- it somehow worked
- **only** RELEASE90 and dev- platform were offered at the start. Why the NB release itself (RELEASE130) was not in the immediate list, I don't understand -- so I filed [NETBEANS-6491](https://issues.apache.org/jira/browse/NETBEANS-6491) probably for myself, but won't work on that immediately; low priority as workaround exist
- the result was the following project order:
```
mavenproject1-app
mavenproject1-branding
mavenproject1-parent
```
which is quite misleading for novice users: how should one know that one must build the reactor in `-parent` ? An engineer with some maven knowledge should have guessed from the error message itself, but still.

- next, I had proxy left & turned on in my `.m2/settings.xml` from the previous network where my laptop was connected. Now, on home connection, the proxy caused Maven **not to be able to download** artifacts. Strange messages at the start (artfiact that must have been released to the Central was not available ?!?), but since this is a 'well-known' .... filed [NETBEANS-6492](https://issues.apache.org/jira/browse/NETBEANS-6492)

After these things, the reactor project built cleanly on JDK11.

However, I was not able to reproduce the exact messages copied in the original report; namely
```
Child module /Users//mavenproject1/branding of /Users//mavenproject1/pom.xml does not exist 
```
Is very strange, as the autogenerated module/subproject is always named "branding", not "branding of ...". @Addlai are you sure that you copied the message exactly ? Also note the duplicated slash after `/Users`.

**Reply by neilcsmith-net** [reply]
Thanks @sdedic Note that issues if filed should be here. We're waiting on infra@ to make JIRA effectively read-only.

**Reply by sdedic** [reply]
> Thanks @sdedic Note that issues if filed should be here. We're waiting on infra@ to make JIRA effectively read-only.

Hm, then perhaps I'll stop filing issues at all - and won't track / plan work based on identified issues - let me say that in (not only) my opinion the GH system simply sucks and I don't find it suitable for managing / planning work items and tracking dependencies between them. I more or less refuse to use it. So if the community wants to have JIRA r/o, it should expect less contributions from my side.

**Reply by neilcsmith-net** [reply]
That decision was made months ago on dev@ now.

**Reply by sdedic** [reply]
> That decision was made months ago on dev@ now.

I know.  My opinion on using GH issue system is constantly the same - back then and now.

I didn't want to step in, as it can be seen as rather inappropriate influence of Oracle on community processes; however contributing issues, or doing basic planning here on JIRA/Github was and is not required for my work. And since I do this 'in my own capacity' all the time, as a 'volunteer', I have also the liberty to NOT do that voluntary work if I am required to use tools that I find inappropriate. That's all.

**Reply by neilcsmith-net** [reply]
I disagree (on influence). That just meant no-one on the PMC spoke in favour of JIRA and stepped up to fix the integration issues. This is going off topic anyway. If there's a need to keep some permissions in JIRA, please raise soon on dev@ before my infra ticket is actioned.

**Reply by mbien** [reply]
tbh i would only file issues if the dev solving the issue != the dev filing the issue. If you want to fix it yourself anyway, just open a draft PR and link some other PRs or issues if needed.

I did that for #3558 for example. I am a freelancer myself, so anything i contribute to NB (or other open source projects) is done outside of "work". So i have the exact same issue as @sdedic - just inverted, I refuse to use JIRA in my free time unless i have to :)

**Reply by sdedic** [reply]
> tbh i would only file issues if the dev solving the issue != the dev filing the issue. If you want to fix it yourself anyway, just open a draft PR and link some other PRs or issues if needed.
> 
I am used to note ideas or identified tasks this way - as this is not a priority, I might or might not (= someone else will) get to it.  In this case, these enhancements were simply byproducts of testing scenario in this topic. Draft PR is (IMHO) not suitable to make such notes; they mess in the PR list with real intentions to code, or prototypes. Anyway I'd suggest not messing this with process discussion further - take it on dev@, if you like though I won't participate (see above).

**Reply by mbien** [reply]
i mean @neilcsmith-net said it already: the discussion happened some time ago on the dev list. It doesn't matter what company you are working for at the moment, if you are involved in he project you have the right to participate in discussions on the dev list.

Saying nothing but then disagreeing with the result is a situation we should avoid. (I probably misunderstood you but it reads to me as if you don't want to discuss it here but also not on the dev list). Lets try to find a solution (somewhere, I don't care where).

For example: we could keep jira open for meta tasks if some devs prefer to use it over gh issues. But how should I suggest that on the dev list if a) i wouldn't use it myself and b) you already said you would not participate in the discussion. This is not going to work.

**Reply by sdedic** [reply]
@mbien please let's focus back on the reporter's issues. Thanks for understanding.

**Reply by neilcsmith-net** [reply]
@sdedic yes please!  To which I actually come back to my original point.  It's pointless opening issues on JIRA for these, particularly when it's about to be switched off.  No-one else is going to pick up on these if they're there.  Issues here and tagged with NB14 at least have a chance of someone else taking a look before we freeze.
